localstack/services/sqs/models.py

```diff
@@ -1018,15 +1014,20 @@ def _put_message(self, message: SqsMessage):
         message_group = self.get_message_group(message.message_group_id)
 
         with self.mutex:
+            previously_empty = message_group.empty()
             # put the message into the group
             message_group.push(message)
 
-            # if an older message becomes visible again in the queue, that message's group becomes visible also.
-            if message.receive_count < 1:
+            # new messages should not make groups visible that are currently inflight
+            if message.receive_count < 1 and message_group in self.inflight_groups:
                 return
+            # if an older message becomes visible again in the queue, that message's group becomes visible also.
             if message_group in self.inflight_groups:
                 self.inflight_groups.remove(message_group)
                 self.message_group_queue.put_nowait(message_group)
+            # if the group was previously empty, it was not yet added back to the queue
+            elif previously_empty:
+                self.message_group_queue.put_nowait(message_group)
```

Comment:
nit: Should we consolidate some of the branches, and make the unhandled case explicit with an exception, or at least a log message?

```python

            if message_group in self.inflight_groups:
                # new messages should not make groups visible that are currently inflight
                if message.receive_count < 1:
                    return

                # if an older message becomes visible again in the queue, that message's group becomes
                # visible also.
                self.inflight_groups.remove(message_group)
                self.message_group_queue.put_nowait(message_group)
            # if the group was previously empty, it was not yet added back to the queue
            elif previously_empty:
                self.message_group_queue.put_nowait(message_group)
            else:
                raise ValueError(
                    f"Message group {message_group.message_group_id} is in an unknown state"
                ) # or LOG.warning(...)
```